File: app/src/main/java/com/icerrate/popularmovies/provider/db/MovieContentProvider.java
package com.icerrate.popularmovies.provider.db;

import android.content.ContentProvider;
import android.content.ContentValues;
import android.content.UriMatcher;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.net.Uri;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import static com.icerrate.popularmovies.provider.db.MovieDBHelper.TABLE_FAVORITE;

/**
 * @author Ivan Cerrate.
 */

public class MovieContentProvider extends ContentProvider {

    private static final UriMatcher uriMatcher = new UriMatcher(UriMatcher.NO_MATCH);

    private MovieDBHelper movieDBHelper;

    @Override
    public boolean onCreate() {
        movieDBHelper = new MovieDBHelper(getContext());
        return true;
    }

    @Nullable
    @Override
    public Cursor query(@NonNull Uri uri, @Nullable String[] projection, @Nullable String selection, @Nullable String[] selectionArgs, @Nullable String sortOrder) {
        final SQLiteDatabase db = movieDBHelper.getReadableDatabase();
        return db.query(TABLE_FAVORITE, projection, selection, selectionArgs, null, null, sortOrder);
    }

    @Nullable
    @Override
    public Uri insert(@NonNull Uri uri, @Nullable ContentValues values) {
        final SQLiteDatabase db = movieDBHelper.getWritableDatabase();
        long id = db.insert(TABLE_FAVORITE, null, values);
        return Uri.parse(TABLE_FAVORITE+"/" + id);
    }

    @Override
    public int delete(@NonNull Uri uri, @Nullable String selection, @Nullable String[] selectionArgs) {
        final SQLiteDatabase db = movieDBHelper.getWritableDatabase();
        return db.delete(TABLE_FAVORITE, selection, selectionArgs);
    }

    @Override
    public int update(@NonNull Uri uri, @Nullable ContentValues values, @Nullable String selection, @Nullable String[] selectionArgs) {
        final SQLiteDatabase db = movieDBHelper.getWritableDatabase();
        String id = uri.getLastPathSegment();
        return db.update(TABLE_FAVORITE, values, "_id=" + id, selectionArgs);
    }

    @Nullable
    @Override
    public String getType(@NonNull Uri uri) {
        return null;
    }
}

File: app/src/main/java/com/icerrate/popularmovies/provider/db/MovieContentProvider.kt
package com.icerrate.popularmovies.provider.db

import android.content.ContentProvider
import android.content.ContentValues
import android.content.UriMatcher
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import android.net.Uri
import com.icerrate.popularmovies.provider.db.MovieDBHelper.Companion.TABLE_FAVORITE

/**
 * @author Ivan Cerrate.
 */
class MovieContentProvider : ContentProvider() {

    companion object {
        private val uriMatcher = UriMatcher(UriMatcher.NO_MATCH)
    }

    private lateinit var movieDBHelper: MovieDBHelper

    override fun onCreate(): Boolean {
        movieDBHelper = MovieDBHelper(context!!)
        return true
    }

    override fun query(
        uri: Uri,
        projection: Array<String>?,
        selection: String?,
        selectionArgs: Array<String>?,
        sortOrder: String?
    ): Cursor? {
        val db: SQLiteDatabase = movieDBHelper.readableDatabase
        return db.query(TABLE_FAVORITE, projection, selection, selectionArgs, null, null, sortOrder)
    }

    override fun insert(uri: Uri, values: ContentValues?): Uri? {
        val db: SQLiteDatabase = movieDBHelper.writableDatabase
        val id = db.insert(TABLE_FAVORITE, null, values)
        return Uri.parse("$TABLE_FAVORITE/$id")
    }

    override fun delete(uri: Uri, selection: String?, selectionArgs: Array<String>?): Int {
        val db: SQLiteDatabase = movieDBHelper.writableDatabase
        return db.delete(TABLE_FAVORITE, selection, selectionArgs)
    }

    override fun update(
        uri: Uri,
        values: ContentValues?,
        selection: String?,
        selectionArgs: Array<String>?
    ): Int {
        val db: SQLiteDatabase = movieDBHelper.writableDatabase
        val id = uri.lastPathSegment
        return db.update(TABLE_FAVORITE, values, "_id=?", arrayOf(id!!))
    }

    override fun getType(uri: Uri): String? {
        return null
    }
}